s3_management/manage.py

```diff
@@ -212,6 +213,23 @@ def normalize_package_version(self: S3IndexType, obj: str) -> str:
     def obj_to_package_name(self, obj: str) -> str:
         return path.basename(obj).split('-', 1)[0]
 
+    def fetch_checksum_from_s3(self, s3_key):
+        s3_key = s3_key.replace("%2B", "+")
+        try:
+            response = CLIENT.get_object_attributes(
+                Bucket=BUCKET,
+                Key=s3_key,
+                ObjectAttributes=['Checksum']
+            )
+            checksum = response['Checksum']['ChecksumSHA256']
```

Comment:
Is there any way to leverage the call in `from_S3` to also include this information?

If not, perhaps this should be done there so the returned objects already have the checksum attached and the downstream code can choose to use it or not?

That would isolate the S3 retrieval code to be in one logical place.

Reply:
> Is there any way to leverage the call in from_S3 to also include this information?

From what I can tell they are different clients/endpoints that serve different purposes.

>If not, perhaps this should be done there so the returned objects already have the checksum attached and the downstream code can choose to use it or not?

From my perspective, the ``from_S3` appears to just be tracking a list of strings which is the object s3_key with `List[str]` so I am not sure it makes sense to tackle a larger refactor just to add all the additional calls to pre-lookup these checksums and store them in memory until their point of use.  

> That would isolate the S3 retrieval code to be in one logical place.

Isn't it already isolated to this one file manage.py where the clients are in scope for the whole file?

Maybe I am being dense and don't see the obvious way to accomplish this without spending a couple more hours on it, plus the more complexity I add, well I have no real way of testing it.  Looking for any guidance on the right exceptions to catch or not catch above, but I think we need to catch something to cover the case that s3 isn't ready or able to provide the hashes.

Reply:
OK, so looks like you should be able to do this in `from_s3` (which is ideal, because we'll want the checksums for metadata files in https://github.com/pytorch/builder/pull/1457 as well:

```python
>>> import boto3
>>> BUCKET=S3.Bucket("<bucketname>")
>>> obj = next(iter(BUCKET.objects.filter(Prefix="<some prefix>")))
>>> response = obj.meta.client.head_object(Bucket=BUCKET.name, Key=obj.key, ChecksumMode="ENABLED")
>>> response["ChecksumSHA256"]
'7Uc3AtM1Dlc4XiQ8A6NkQdV9JMU4liPtz7wOd+RRCTE='
```

And then for an object without a checksum:
```python
>>> >>> response = obj.meta.client.head_object(Bucket=BUCKET.name, Key=obj.key, ChecksumMode="ENABLED")
>>> list(response)
['ResponseMetadata', 'AcceptRanges', 'LastModified', 'ContentLength', 'ETag', 'ContentType', 'ServerSideEncryption', 'Metadata']
```

Reply:
Resolved.

Reply:
Build is failing with:
```botocore.exceptions.ParamValidationError: Parameter validation failed:
Unknown parameter in input: "ChecksumMode", must be one of: Bucket, IfMatch, IfModifiedSince, IfNoneMatch, IfUnmodifiedSince, Key, Range, VersionId, SSECustomerAlgorithm, SSECustomerKey, SSECustomerKeyMD5, RequestPayer, PartNumber```

Reply:
https://github.com/pytorch/builder/pull/1548

Reply:
Tagging @malfet again, as these inner comments are notoriously hard to discover. Hope it's no big deal 😇